### ISSUE TITLE
ConfigurationError: Experiment specified GPU device 2 but there are only 1 devices  available

### ISSUE THREAD

**Issue by tingkai-zhang** [allenai/allennlp#2590]
**Describe the bug**

The GPU 0 and 1 on server are occupied fully . 

So I want to put my model on device 2.

I set `cuda_device` as 2 in config file and tried to set `CUDA_VISIBLE_DEVICES`  as 2 either by adding `CUDA_VISIBLE_DEVICES` at the beginning of training command  or add line `os.environ["CUDA_VISIBLE_DEVICES"] = "2"` in my model file.

It will gives the error `Experiment specified GPU device 2 but there are only 1 devices  available.` from [this line](https://github.com/allenai/allennlp/blob/9437b614dbc518b3b532bbaa201446dc454fc6e2/allennlp/common/checks.py#L61).

Why number of device(1) has to be greater than device id(2)? 
(I have seen the comment in #1090 

> If CUDA_VISIBLE_DEVICES is set to a single value of the corresponding GPU id, then you should just use self.config.gpu_id = 0, and it will use the right GPU. 

so i also tried to set the `cuda_device` of `trainer` as 0 in config file.  But it will gave OOM error at the beginning of a epoch)

What is possible work around for this problem?

**To Reproduce**
 Set `cuda_device` as 2 in config file and set `CUDA_VISIBLE_DEVICES`  as 2 in model file.

**Expected behavior**
Model will be trained on device 2 and no GPU leakage to GPU 0.

**System (please complete the following information):**
 - OS: Linux
 - AllenNLP version:  I installed from master

**Comment by epwalsh** [comment]
I believe setting `CUDA_VISIBLE_DEVICES` causes CUDA to re-enumerate the devices available. So in your case device 2 would become device 0.

**Comment by tingkai-zhang** [comment]
Hi, @epwalsh ! Thanks for replying!

Do you mean I should set `cuda_device` of trainer as 0 in config file? 
I did.

I set `CUDA_VISIBLE_DEVICES` in model file and 
> I also tried to set the cuda_device of trainer as 0 in config file. But it will give OOM error)

**Comment by epwalsh** [comment]
Hi @tingkai-zhang, if the OOM error occurs during an epoch of training you could try using a smaller batch size. If your inputs or outputs are sequences you could also try truncating them.

**Comment by tingkai-zhang** [comment]
@epwalsh 
Thanks but that's not the problem root.

I trained exactly the same model on one GPU and it has no OOM error.

It is because pytorch will create some context on GPU 0 and it will give OOM error when GPU 0 is fully occupied.
That's why I set `cuda_device` as 2 and `CUDA_VISIBLE_DEVICES` as 2.

**Comment by tingkai-zhang** [comment]
I spent some time investigating it. The reason is not about AllenNLP. So I close this issue now. Thank you!

**Comment by doudouzqx** [comment]
Has your problem been solved? I also encountered this situation when used multiple GPUs.

**Comment by tingkai-zhang** [comment]
Sorry, I forgot how I solved the problem but I remember that's not the issue of AllenNLP in my case. @doudouzqx

**Comment by gohjiayi** [comment]
I was faced with the following error although I have set `export CUDA_VISIBLE_DEVICES=0,1,2,3` on my CLI. Apparently it meant that I only had 1 GPU available (which was false as I had 4).

```
allennlp.common.checks.ConfigurationError: Experiment specified GPU device 1 but there are only 1 devices  available.
```

I was able to resolve this by editing the parameters .jsonnet file (available from [this tutorial here](https://towardsdatascience.com/pytorch-elmo-844d2391a0b2#b484)). This is part of the original parameters.

```
...
"distributed": {
    "cuda_devices": if NUM_GPUS > 1 then std.range(0, NUM_GPUS - 1) else 0,
  },
...
```

I simply edited to `"cuda_devices": [0,1,2,3]` (as I had 4 GPUs available). You are able to change the `"cuda_devices"` parameter under `"trainer"` too instead of `"distributed"` like my case. Check out allennlp tutorials for more info on fields within the parameter files.
